sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/tests/TimeSeriesInsightsQueryTests.cs

```diff
@@ -0,0 +1,228 @@
+                    await foreach (QueryResultPage eventPage in queryEventsPages)
+                    {
+                        eventPage.Timestamps.Should().HaveCount(2);
+                        eventPage.Properties.Should().NotBeEmpty().And.HaveCount(2);
+                        eventPage.Properties.First().Should().NotBeNull();
+                        eventPage.Properties.First().Name.Should().Be("Temperature");
+                        eventPage.Properties[1].Name.Should().Be("building");
+                    }
+                    return null;
+                }, MaxNumberOfRetries, s_retryDelay);
+
+                // Query for the two events with a filter, but only take 1
+                queryRequestOptions.MaximumNumberOfEvents = 1;
+                AsyncPageable<QueryResultPage> queryEventsPagesWithFilter = tsiClient.QueryEventsAsync(tsiId, startTime, endTime, queryRequestOptions);
+                await foreach (QueryResultPage eventPage in queryEventsPagesWithFilter)
+                {
+                    eventPage.Timestamps.Should().HaveCount(1);
+                    eventPage.Properties.Should().NotBeEmpty().And.HaveCount(2);
+                    eventPage.Properties.First().Should().NotBeNull();
+                    eventPage.Properties.First().Name.Should().Be("Temperature");
+                    eventPage.Properties[1].Name.Should().Be("building");
+                }
+
+                // Query for all the events using a timespan
+                AsyncPageable<QueryResultPage> queryEventsPagesWithTimespan = tsiClient.QueryEventsAsync(tsiId, TimeSpan.FromMinutes(10));
+                await foreach (QueryResultPage eventPage in queryEventsPagesWithTimespan)
+                {
+                    eventPage.Timestamps.Should().HaveCount(6);
+                    eventPage.Timestamps.Should().OnlyContain(timeStamp => timeStamp >= startTime).And.OnlyContain(timeStamp => timeStamp <= endTime);
+                    eventPage.Properties.Should().NotBeEmpty();
+                    eventPage.Properties.First().Should().NotBeNull();
+                }
+
+                // Query for all the events using a timespan and an old end date
+                AsyncPageable<QueryResultPage> queryEventsPagesWithOldTimespan = tsiClient
+                    .QueryEventsAsync(tsiId, TimeSpan.FromMinutes(1), DateTimeOffset.UtcNow.AddDays(-1));
+                await foreach (QueryResultPage eventPage in queryEventsPagesWithOldTimespan)
+                {
+                    eventPage.Timestamps.Should().HaveCount(0);
+                }
+            }
+            finally
+            {
+                deviceClient?.Dispose();
+            }
+        }
+
+        private async Task SendEventsToHubAsync(DeviceClient client, TimeSeriesId tsiId, TimeSeriesIdProperty[] timeSeriesIdProperties, int numberOfEventsToSend)
+        {
+            IDictionary<string, object> messageBase = BuildMessageBase(timeSeriesIdProperties, tsiId);
```

Comment:
I fear this will not work for recorded tests. Have you thought about that?

There's also query timeframes. Running it now, the results will be recent, but when replaying a recorded tests the data points will be further back.

Reply:
Sorry, because track1 device client SDK won't support recording, even if using HTTP.

Reply:
I used Recording.UtcNow to be able to account for Record/Playback